compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -1001,7 +1001,20 @@ where
         kind: MemoryKind<M::MemoryKind>,
         mutbl: Mutability,
     ) -> InterpResult<'tcx, MPlaceTy<'tcx, M::Provenance>> {
-        let ptr = self.allocate_bytes_ptr(str.as_bytes(), Align::ONE, kind, mutbl)?;
+        let tcx = self.tcx.tcx;
+
+        // Use cache for immutable strings.
+        let ptr = if mutbl.is_not()
+            && matches!(kind, MemoryKind::CallerLocation | MemoryKind::Machine(_))
```

Comment:
Why is this only done for some `kind`?

Reply:
My thinking is that `MemoryKind::Stack` can get [deallocated](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_const_eval/interpret/eval_context.rs.html#965) while global memory [shouldn't](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_const_eval/interpret/memory.rs.html#309) be deallocated.

Reply:
Immutable memory can never be deallocated. And anyway memory in the global `tcx` map can also never be deallocated.

Reply:
We could just ICE on other memory kinds so we'll know when someone actually tries to use them

Reply:
> Immutable memory can already never be deallocated.

Digging a bit by myself, I found that [mutability check](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_const_eval/interpret/memory.rs.html#337) in `deallocate_ptr` is done after global memory check.
So `deallocate_ptr` ensures that global memory and local `InterpCx`'s immutable memory are not deallocated.

If someone uses `allocate_str` to allocate immutable memory and then deallocate it, it would trigger an global memory deallocation error.
But they shouldn't deallocate immutable memory anyway, global or not. So it's all good.

`MemoryKind` is irrelevant here and stack variables with heap allocations have no need to use `allocate_str`.